stan/math/prim/prob/cauchy_lpdf.hpp

```diff
@@ -37,13 +38,17 @@ template <bool propto, typename T_y, typename T_loc, typename T_scale>
 return_type_t<T_y, T_loc, T_scale> cauchy_lpdf(const T_y& y, const T_loc& mu,
                                                const T_scale& sigma) {
   using T_partials_return = partials_return_t<T_y, T_loc, T_scale>;
+  using T_partials_array = Eigen::Array<T_partials_return, Eigen::Dynamic, 1>;
   using std::log;
+  using T_y_ref = ref_type_if_t<!is_constant<T_y>::value, T_y>;
```

Comment:
What is the logic here? This is the type that will be responsible for .eval() ing the input right?

Reply:
Right, we are only evaling this if it is not constant. The reason being is that `y_ref` seems to be used twice. Once in construction of `operands_and_partials` and once in creation of `y_col`. However, prim implementation of `operands_and_partials` does not actually do anything with the arguemnts. So if y is double it is actually only used once and we don't need to eval it.

Reply:
Since it will get used at least once, it would always be safe to eval it though. So we can get rid of the conditional and it'll be the same.

Reply:
Yeah it is safe. However I prefere not to for performance reasons. If we avaluate multiple operations in same expression we need to load the data into cache just once. If we evaluate each operation separately we need to load them every time.

Reply:
I was looking at `ref_type_if_t` to see what it did and it looks like it returns a `plain_type` if it doesn't do a reference type.

That will induce a copy anyway: https://mc-stan.org/math/d2/d08/ref__type_8hpp_source.html

Reply:
Assuming `T_y` is `MatrixXd` this is `ref_type_if_t<false, MatrixXd>`, so its results is set to `T_optionally_ref`, which is `const MatrixXd&`. No copy as I see it.

Reply:
Right but that will still cause an eval. In which case we're evaling this regardless of it is constant or not and so:

```
using T_y_ref = ref_type_t<T_y>;
```

will achieve the same thing in the end as:

```
using T_y_ref = ref_type_if_t<!is_constant<T_y>::value, T_y>;
```

Reply:
@bbbales2 I hope this is clear now. Or is it ... ?

Reply:
Yeah it's clear. I delayed cause this (and the other lpdf pulls) are major changes to lpdf and I got nervous and wanted at least https://github.com/stan-dev/math/pull/1989 to get in.

Until that is in there's at least a couple weaknesses in the testing framework that let bugs through before (though only one of those is with pmfs/pdfs - https://github.com/stan-dev/math/issues/1861).

I didn't get around to breaking that pull up Friday.